protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+Arguably this is the Platforms-Protocol team split, but things can be fuzzy between them.
+
+With both platforms and protocol teams we can align on what improvements we need,
+and implement them collaboratively.
+
+# Problem Statement + Context
+
+<!-- Describe the specific problem that the document is seeking to address as well
+as information needed to understand the problem and design space.
+If more information is needed on the costs of the problem,
+this is a good place to that information. -->
+
+As outlined above, there are different areas where the stack is expanding,
+and where testing needs to expand with it.
+
+What we currently have may be sufficient for a while,
+but the pressure of a "test ceiling" does also stall development.
+E.g. more `op-e2e` bloat -> more `op-e2e` flakes -> less confidence in `op-e2e` -> fewer features / changes.
+
+*Some pressure* may be good: complexity has to stop somewhere.
+But ideally this comes as a design-choice, not a pain/delay in development.
+
+The sections below review what we have today, what challenges we have,
+and what ideas we have, for each of the growth domains.
+
+### Growth of chains
+
+With a greater number of chains, we will need more test deployments,
+and more monitoring.
+
+Today:
+- Superchain-registry [validation checks](https://github.com/ethereum-optimism/superchain-registry/tree/main/validation)
+- Test nodes on each network, with alerts
+- [pessimism](https://github.com/base-org/pessimism) monitoring (deprecated)
+- [monitorism](https://github.com/ethereum-optimism/monitorism)
+
+Challenges:
+- Test nodes can be unstable: lack of peers can result in missed blocks.
+- Test nodes can elevate costs: not everything has to run on the same tier hardware as sequencers.
+  There is a economies-of-scale thing to investigate.
+- Monitorism may need to be deployed to more chains.
+
+
+### Growth of features
+
+"Mo' features, mo' problems." Or really, more edge-cases to validate.
+We need the test-infrastructure to efficiently express how things can go wrong,
+and what we expect to happen in each case.
+
+Today:
+- [`op-e2e` system tests](https://github.com/ethereum-optimism/optimism/tree/develop/op-e2e/system)
+- [`op-e2e` action tests](https://github.com/ethereum-optimism/optimism/tree/develop/op-e2e/actions)
+- [smart contracts tests](https://github.com/ethereum-optimism/optimism/tree/develop/packages/contracts-bedrock/test)
+- [op-chain-ops upgrade checks](https://github.com/ethereum-optimism/optimism/tree/develop/op-chain-ops/cmd)
+- [superchain-ops](https://github.com/ethereum-optimism/superchain-ops/) checks
+- Interop tests (see [interop section](#interoperability))
+
+Challenges:
+- System tests are prone to flakes: running N systems,
+  in a resource-constrained environment, with concurrent work / timers / etc., tends to miss assertions due to processing stalls.
+- `op-e2e` tests are not portable to alternative client implementations.
```

Comment:
Big +1 on this. The tests themselves should be decoupled from the clients they run against, so we can build matrix-style tests against multiple clients.

Reply:
But is op-e2e the right framework for that? The vast majority of op-e2e tests aren't testing the EL. Many are testing op-node derivation, many are testing op-proposer, op-batcher, op-challenger, dispute games etc none of which have alternate implementations.

We do need a multi client testsuite, but I'm not sure having to pull everything into op-e2e is the right answer. We don't want op-reth, op-nethermind, op-besu, op-whatever-I-just-created to be a dependency of the monorepo for example. Asterisc pulls the monorepo in so it can run e2e tests in their repo and it's not a particularly clean solution - it works but we break it by making changes to op-e2e pretty regularly.

I'm a really big fan of being able to export a test suite like the L1 reference tests do.  Individual clients will still have their own e2e tests in addition for things that aren't consensus critical but the reference tests define the shared set of tests for consensus compatibility and are designed to be implementation and test framework agnostic and easy to pull a release into whatever repo needs to run them.